01 API Tutorials/04 Using Options in QuantConnect/04 Select Contracts.html

```diff
@@ -65,11 +65,11 @@
 <pre class="python">def OnData(self,slice):
     for i in slice.OptionChains:
         if i.Key != self.symbol: continue
-	optionchain = i.Value
-	self.Log("underlying price:" + str(optionchain.Underlying.Price))
-	df = pd.DataFrame([[x.Right,float(x.Strike),x.Expiry,float(x.BidPrice),float(x.AskPrice)] for x in optionchain],
-			   index=[x.Symbol.Value for x in optionchain],
-			   columns=['type(call 0, put 1)', 'strike', 'expiry', 'ask price', 'bid price'])
+	      optionchain = i.Value
+	      self.Log("underlying price:" + str(optionchain.Underlying.Price))
+        df = pd.DataFrame([[x.Right,float(x.Strike),x.Expiry,float(x.BidPrice),float(x.AskPrice)] for x in optionchain],
+        index=[x.Symbol.Value for x in optionchain],
+        columns=['type(call 0, put 1)', 'strike', 'expiry', 'ask price', 'bid price'])
         self.Log(str(df))
```

Comment:
We could take the opportunity to also rename `i` for `kvp` since `slice.OptionChains` is a `KeyValuePair`.
There is no need to convert the prices to `float` and we should keep `index` and `column` aligned to the method parenthesis:
```python
        for kvp in slice.OptionChains:
            if kvp.Key != self.symbol:
                continue
            optionchain = kvp.Value

            df = pd.DataFrame([[x.Right, x.Strike, x.Expiry, x.BidPrice, x.AskPrice] for x in optionchain],
                              index=[x.Symbol.Value for x in optionchain],
                              columns=['type(call 0, put 1)', 'strike', 'expiry', 'ask price', 'bid price'])

            self.Log(f"Underlying price: {optionchain.Underlying.Price}\n{df}")
```
I also suggest merging the Log calls.

Reply:
True, that makes sense.